tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -901,16 +901,27 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     const std::vector<HloInstruction *> gemm_users = existing_gemm->users();
     HloInstruction *reduce_damax = nullptr;
     if (gemm_users.size() == 2) {
+      // In the presence of a ReLU activation, the abs instruction is elided
+      // since abs(ReLU(x)) = ReLU(x).
+      TF_ASSIGN_OR_RETURN(auto config,
+                          existing_gemm->backend_config<GemmBackendConfig>());
+      bool has_relu_epilogue =
+          (config.epilogue() == GemmBackendConfig::BIAS_RELU ||
+           config.epilogue() == GemmBackendConfig::RELU);
       for (int i = 0; i < gemm_users.size(); ++i) {
-        if (gemm_users[i]->opcode() == HloOpcode::kAbs &&
-            gemm_users[i]->users().size() == 1 &&
-            gemm_users[i]->users()[0]->opcode() == HloOpcode::kReduce &&
-            gemm_users[i]->users()[0]->operands().size() == 2 &&
-            gemm_users[i]->users()[0]->operand(1)->opcode() ==
-                HloOpcode::kConstant &&
-            ShapeUtil::IsScalar(
-                gemm_users[i]->users()[0]->operand(1)->shape())) {
-          HloInstruction *reduce = gemm_users[i]->users()[0];
+        if (!has_relu_epilogue && gemm_users[i]->users().empty()) {
+          continue;
+        }
+        HloInstruction *maybe_reduce =
+            has_relu_epilogue ? gemm_users[i] : gemm_users[i]->users()[0];
```

Comment:
This logic is somewhat hard to follow. I would replace these lines with something like

```c++
HloInstruction* maybe_reduce = nullptr;
if (gemm_users[i]->opcode() == HloOpcode::kAbs) {
  if (gemm_users[i]->users().size() != 1) continue;
  maybe_reduce = gemm_users[i]->users()[0];
} else {
  // If there is no Abs instruction, we require a relu epilogue to ensure all values are nonnegative.
  if (!has_relu_epilogue) continue;
  maybe_reduce = gemm_users[i];
}
```

This allows you to remove the `has_relu_epilogue` check and the `abs` check below as well. This logic also allows the Abs to appear with a Relu, which I didn't intend but doesn't hurt.